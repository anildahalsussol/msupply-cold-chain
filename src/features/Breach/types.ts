import { TemperatureBreach } from '~common/services/Database';

export interface AcknowledgeBreachSliceState {
  acknowledgingSensorId: string;
  acknowledging: boolean;
  updating: boolean;
  fetching: boolean;
  unacknowledged: TemperatureBreach[];
}

export interface FetchUnacknowledgedSuccessPayload {
  breaches: TemperatureBreach[];
}

export interface SensorIDPayload {
  sensorId: string;
}
