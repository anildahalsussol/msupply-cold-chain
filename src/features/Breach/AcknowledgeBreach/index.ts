export {
  AcknowledgeBreachAction,
  AcknowledgeBreachReducer,
  AcknowledgeBreachSaga,
  AcknowledgeBreachSelector,
} from './AcknowledgeBreachSlice';

export { AcknowledgeBreachManager } from './AcknowledgeBreachManager';
