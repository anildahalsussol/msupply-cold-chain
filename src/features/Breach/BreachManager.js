import moment from 'moment';
import { Not, IsNull, MoreThan, Equal } from 'typeorm/browser';

import { ENTITIES } from '~constants';

const CUMULATIVE_EXPOSURE = `
select *, sum(logInterval), max(temperature) maximumTemperature,
sum(logInterval) duration,
min(temperature) minimumTemperature,
case when min(temperature) >= hotCumulativeMinThreshold and sum(logInterval) * 1000 >= hotCumulativeDuration then 1 else 0 end as isHotCumulative,
case when max(temperature) <= coldCumulativeMaxThreshold and sum(logInterval) * 1000 >= coldCumulativeDuration then 1 else 0 end as isColdCumulative
from (
select temperature,
logInterval,
(select maximumTemperature from temperaturebreachconfiguration where id = 'HOT_CUMULATIVE') as hotCumulativeMaxThreshold,
(select minimumTemperature from temperaturebreachconfiguration where id = 'HOT_CUMULATIVE') as hotCumulativeMinThreshold,
(select duration from temperaturebreachconfiguration where id = 'HOT_CUMULATIVE') as hotCumulativeDuration,
(select maximumTemperature from temperaturebreachconfiguration where id = 'COLD_CUMULATIVE') as coldCumulativeMaxThreshold,
(select minimumTemperature from temperaturebreachconfiguration where id = 'COLD_CUMULATIVE') as coldCumulativeMinThreshold,
(select duration from temperaturebreachconfiguration where id = 'COLD_CUMULATIVE') as coldCumulativeDuration
from temperaturelog
where ((temperature between hotCumulativeMinThreshold and hotCumulativeMaxThreshold) 
or (temperature between coldCumulativeMinThreshold and coldCumulativeMaxThreshold))
and (timestamp >= ? and timestamp <= ?)
and sensorId = ?
)
group by temperature >= hotCumulativeMinThreshold
`;

const REPORT = `
select (select "Continuous") "Breach Type",
tbc.description "Breach Name",
datetime(startTimestamp, "unixepoch", "localtime") "Start date",
coalesce(datetime(endTimestamp, "unixepoch", "localtime"), datetime("now", "localtime")) as "End date",
(coalesce(endTimestamp, strftime("%s", "now")) - startTimestamp) / 60 "Exposure Duration (minutes)",
max(temperature) as "Max Temp",
min(temperature) as "Min Temp"
from temperaturebreach tb
join temperaturelog tl on tl.temperatureBreachId = tb.id
join temperaturebreachconfiguration tbc on tbc.id = tb.temperatureBreachConfigurationId
where tb.sensorId = ?
group by tb.id
`;

export class BreachManager {
  constructor(databaseService, utils) {
    this.databaseService = databaseService;
    this.utils = utils;
  }

  getAll = async () => {
    return this.databaseService.getAll(ENTITIES.TEMPERATURE_BREACH);
  };

  getLogsForBreach = async breachId => {
    return this.databaseService.queryWith(ENTITIES.TEMPERATURE_LOG, {
      temperatureBreachId: breachId,
    });
  };

  getCumulativeExposure = async (from, to, sensorId) => {
    const manager = await this.databaseService.getEntityManager();
    const result = await manager.query(CUMULATIVE_EXPOSURE, [from, to, sensorId]);

    const lookup = result.reduce((acc, value) => {
      if (value.isHotCumulative) return { ...acc, hotCumulative: value };
      if (value.isColdCumulative) return { ...acc, coldCumulative: value };
      return acc;
    }, {});

    return lookup;
  };

  closeBreach = (temperatureBreach, time) => {
    // eslint-disable-next-line no-param-reassign
    temperatureBreach.endTimestamp = time;
    return temperatureBreach;
  };

  createBreach = (sensor, temperatureBreachConfiguration, startTimestamp) => {
    const { id: sensorId } = sensor;
    const { id: temperatureBreachConfigurationId } = temperatureBreachConfiguration;
    const id = this.utils.uuid();

    return {
      id,
      sensorId,
      temperatureBreachConfigurationId,
      temperatureBreachConfiguration,
      startTimestamp,
    };
  };

  willCreateBreach = (config, logs) => {
    if (!logs.length) return false;
    const { minimumTemperature, maximumTemperature, duration } = config;
    const { timestamp: endTimestamp } = logs[logs.length - 1];
    const { timestamp: startTimestamp } = logs[0];
    const logsDuration = endTimestamp - startTimestamp;
    if (logsDuration < duration / 1000) return false;
    const temperaturesWithinBounds = logs.every(log => {
      const { temperature } = log;
      return temperature <= maximumTemperature && temperature >= minimumTemperature;
    });
    return temperaturesWithinBounds;
  };

  willCreateBreachFromConfigs = (configs, logs) => {
    const configToCreateBreachFrom = configs.find(config => this.willCreateBreach(config, logs));
    return [!!configToCreateBreachFrom, configToCreateBreachFrom];
  };

  addLogToBreach = (breach, log) => {
    const { id: temperatureBreachId } = breach;
    return { ...log, temperatureBreachId };
  };

  willContinueBreach = (breach, log) => {
    if (!breach) return false;
    const { temperatureBreachConfiguration } = breach;
    const { maximumTemperature, minimumTemperature } = temperatureBreachConfiguration;
    const { temperature } = log;

    return temperature >= minimumTemperature && temperature <= maximumTemperature;
  };

  willCloseBreach = (breach, log) => {
    if (!(breach && log)) return false;
    const { temperatureBreachConfiguration } = breach;
    const { maximumTemperature, minimumTemperature } = temperatureBreachConfiguration;
    const { temperature } = log;
    return !(temperature >= minimumTemperature && temperature <= maximumTemperature);
  };

  couldBeInBreach = (log, configs) => {
    return configs.some(
      config =>
        log.temperature >= config.minimumTemperature && log.temperature <= config.maximumTemperature
    );
  };

  createBreaches = (sensor, logs, configs, breach) => {
    let stack = [];
    let currentBreach = breach;

    const breaches = breach && !breach.endTimestamp ? [breach] : [];
    const temperatureLogs = [];

    logs.forEach(log => {
      const couldBeInBreach = this.couldBeInBreach(log, configs);
      const willCloseBreach = this.willCloseBreach(currentBreach, log);
      const willContinueBreach = this.willContinueBreach(currentBreach, log);

      if (willCloseBreach) {
        this.closeBreach(currentBreach, log.timestamp);
        currentBreach = null;
        stack = [];
      }

      if (couldBeInBreach) stack.push(log);
      else stack = [];

      if (willContinueBreach) {
        const updatedLog = this.addLogToBreach(currentBreach, log);
        temperatureLogs.push(updatedLog);
      }

      if (!willContinueBreach) {
        const [willCreateBreach, config] = this.willCreateBreachFromConfigs(configs, stack);
        if (willCreateBreach) {
          const newBreach = this.createBreach(sensor, config, stack[0].timestamp);
          currentBreach = newBreach;
          breaches.push(newBreach);
          const updatedLogs = stack.map(l => this.addLogToBreach(newBreach, l));
          updatedLogs.forEach(ul => temperatureLogs.push(ul));
        }
      }
    });

    const updatedBreaches = breaches.map(
      ({ id, startTimestamp, endTimestamp, temperatureBreachConfigurationId, sensorId }) => ({
        id,
        startTimestamp,
        endTimestamp,
        temperatureBreachConfigurationId,
        sensorId,
      })
    );
    return [updatedBreaches, temperatureLogs];
  };

  getMostRecentBreach = async sensorId => {
    const [mostRecent] = await this.databaseService.queryWith(ENTITIES.TEMPERATURE_BREACH, {
      where: { sensorId },
      order: { startTimestamp: 'DESC' },
      take: 1,
    });

    return mostRecent;
  };

  getMostRecentBreachLog = async sensorId => {
    const [mostRecentBreachLog] = await this.databaseService.queryWith(ENTITIES.TEMPERATURE_LOG, {
      where: { sensorId, temperatureBreachId: Not(IsNull()) },
      order: { timestamp: 'DESC' },
      take: 1,
    });

    return mostRecentBreachLog;
  };

  createBreachesFrom = async sensorId => {
    const { timestamp = moment(0).unix() } = (await this.getMostRecentBreachLog(sensorId)) ?? {};
    return timestamp;
  };

  getLogsToCheck = async sensorId => {
    const timeToCheckFrom = await this.createBreachesFrom(sensorId);

    return this.databaseService.queryWith(ENTITIES.TEMPERATURE_LOG, {
      where: { sensorId, timestamp: MoreThan(timeToCheckFrom) },
      order: { timestamp: 'ASC' },
    });
  };

  updateBreaches = async (breaches, temperatureLogs) => {
    // eslint-disable-next-line no-param-reassign

    const updatedBreaches = await this.databaseService.upsert(
      ENTITIES.TEMPERATURE_BREACH,
      breaches
    );

    const mapped = temperatureLogs.map(({ id, temperature, timestamp }) => ({
      id,
      temperature,
      timestamp,
    }));

    // TODO: SQLite playing funny bugger games when inserting with a FK straight after
    // creating with a FK
    await this.databaseService.upsert(ENTITIES.TEMPERATURE_LOG, mapped);
    const updatedLogs = await this.databaseService.upsert(
      ENTITIES.TEMPERATURE_LOG,
      temperatureLogs
    );

    return [updatedBreaches, updatedLogs];
  };

  getBreachConfigs = async () => {
    const configs = this.databaseService.queryWith(ENTITIES.TEMPERATURE_BREACH_CONFIGURATION, {
      where: [{ id: Equal('HOT_BREACH') }, { id: Equal('COLD_BREACH') }],
    });

    return configs;
  };

  getBreachReport = async id => {
    const manager = await this.databaseService.getEntityManager();
    return manager.query(REPORT, [id]);
  };
}
