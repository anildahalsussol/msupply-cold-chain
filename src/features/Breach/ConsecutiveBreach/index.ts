export {
  ConsecutiveBreachAction,
  ConsecutiveBreachReducer,
  ConsecutiveBreachSaga,
  ConsecutiveBreachSelector,
} from './ConsecutiveBreachSlice';

export { ConsecutiveBreachManager } from './ConsecutiveBreachManager';
