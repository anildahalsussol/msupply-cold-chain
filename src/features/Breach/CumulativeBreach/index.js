export {
  CumulativeBreachAction,
  CumulativeBreachReducer,
  CumulativeBreachSaga,
  CumulativeBreachSelector,
} from './CumulativeBreachSlice';

export { CumulativeBreachManager } from './CumulativeBreachManager';
