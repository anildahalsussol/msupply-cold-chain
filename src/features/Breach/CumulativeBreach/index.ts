export {
  CumulativeBreachAction,
  CumulativeBreachReducer,
  CumulativeBreachSaga,
  CumulativeBreachSelector,
} from './CumulativeBreachSlice';

export { CumulativeBreachLookup, CumulativeBreachManager } from './CumulativeBreachManager';
