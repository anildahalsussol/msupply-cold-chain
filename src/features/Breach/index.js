export { BreachManager } from './BreachManager';
export { BreachAction, BreachReducer, BreachSelector, BreachSaga } from './BreachSlice';
