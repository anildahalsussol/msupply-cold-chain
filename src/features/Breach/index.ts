export { BreachReducer, BreachSaga } from './BreachSlice';

export {
  CumulativeBreachAction,
  CumulativeBreachSelector,
  CumulativeBreachSaga,
  CumulativeBreachReducer,
  CumulativeBreachManager,
  CumulativeBreachLookup,
} from './CumulativeBreach';

export {
  ConsecutiveBreachAction,
  ConsecutiveBreachSelector,
  ConsecutiveBreachSaga,
  ConsecutiveBreachReducer,
  ConsecutiveBreachManager,
} from './ConsecutiveBreach';

export {
  AcknowledgeBreachAction,
  AcknowledgeBreachSelector,
  AcknowledgeBreachSaga,
  AcknowledgeBreachReducer,
  AcknowledgeBreachManager,
} from './AcknowledgeBreach';
