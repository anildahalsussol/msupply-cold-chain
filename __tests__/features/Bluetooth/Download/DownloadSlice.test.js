import { expectSaga } from 'redux-saga-test-plan';
import { getContext, put } from 'redux-saga/effects';
import {
  DownloadReducer,
  DownloadInitialState,
  DownloadAction,
  DownloadSaga,
} from '~features/Bluetooth';

import { DEPENDENCY } from '~constants';
import { ConsecutiveBreachAction, CumulativeBreachAction } from '~features/Breach';

describe('DownloadReducer', () => {
  it('passiveDownloadingState', () => {
    const state = DownloadReducer(DownloadInitialState, DownloadAction.passiveDownloadingStart());
    expect(state).toEqual({ downloadingById: {}, passiveDownloadEnabled: false, enabled: true });
  });
  it('passiveDownloadingStop', () => {
    const state = DownloadReducer(DownloadInitialState, DownloadAction.passiveDownloadingStop());
    expect(state).toEqual({ downloadingById: {}, passiveDownloadEnabled: false, enabled: false });
  });
  it('downloadStart', () => {
    const state = DownloadReducer(DownloadInitialState, DownloadAction.downloadStart('1'));
    expect(state).toEqual({
      downloadingById: { 1: true },
      passiveDownloadEnabled: false,
      enabled: false,
    });
  });
  it('downloadStart', () => {
    const state = DownloadReducer(DownloadInitialState, DownloadAction.downloadComplete('1'));
    expect(state).toEqual({
      downloadingById: { 1: false },
      passiveDownloadEnabled: false,
      enabled: false,
    });
  });
});

describe('DownloadActions', () => {
  it('', () => {
    const actions = Object.values(DownloadAction).map(actionCreator => actionCreator());
    expect(actions).toMatchSnapshot();
  });
});

describe('DownloadSaga', () => {
  it('tryManualDownloadForSensor: correctly puts actions when a download is possible', () => {
    const mockSensor = { id: '1', macAddress: 'AA:BB:CC:DD:EE:FF', batteryLevel: 90 };
    const sensorManager = {
      getSensorById: () => mockSensor,
      getCanDownload: () => [true],
      getMostRecentLogTime: () => 0,
    };
    const btService = {
      downloadLogsWithRetries: () => [{ temperature: 1 }],
      updateLogIntervalWithRetries: () => {},
    };
    const downloadManager = {
      calculateNumberOfLogsToSave: () => 1,
      createLogs: () => [{ id: '1' }],
      saveLogs: () => [{ id: '1' }],
    };
    const getSensorManager = () => [btService, sensorManager, downloadManager];
    const depsLocator = { get: getSensorManager };

    return expectSaga(DownloadSaga.tryDownloadForSensor, { payload: { sensorId: '1' } })
      .provide([[getContext(DEPENDENCY.LOCATOR), depsLocator]])
      .run()
      .then(({ effects }) => {
        expect(effects.put.length).toBe(5);
        expect(effects.put[0]).toEqual(put(DownloadAction.downloadStart(mockSensor.id)));
        expect(effects.put[1]).toEqual(put(ConsecutiveBreachAction.create(mockSensor)));
        expect(effects.put[2]).toEqual(put(DownloadAction.passiveDownloadForSensorSuccess()));
        expect(effects.put[3]).toEqual(
          put(CumulativeBreachAction.fetchListForSensor(mockSensor.id))
        );
      });
  });
  it('tryManualDownloadForSensor: when cannot download, put fail and complete actions', () => {
    const mockSensor = { id: '1', macAddress: 'AA:BB:CC:DD:EE:FF', batteryLevel: 90 };
    const sensorManager = { getSensorById: () => mockSensor, getCanDownload: () => [false] };
    const btService = {};
    const downloadManager = {};
    const getSensorManager = () => [btService, sensorManager, downloadManager];
    const depsLocator = { get: getSensorManager };

    return expectSaga(DownloadSaga.tryDownloadForSensor, { payload: { sensorId: '1' } })
      .provide([[getContext(DEPENDENCY.LOCATOR), depsLocator]])
      .run()
      .then(result => {
        const { effects } = result;

        expect(effects.put.length).toBe(2);
        expect(effects.put[0]).toEqual(put(DownloadAction.passiveDownloadForSensorFail()));
        expect(effects.put[1]).toEqual(put(DownloadAction.downloadComplete('1')));
      });
  });
  it('tryManualDownloadForSensor: correctly sets the store state when cannot download', () => {
    const mockSensor = { id: '1', macAddress: 'AA:BB:CC:DD:EE:FF', batteryLevel: 90 };
    const sensorManager = { getSensorById: () => mockSensor, getCanDownload: () => [false] };
    const btService = {};
    const downloadManager = {};
    const getSensorManager = () => [btService, sensorManager, downloadManager];
    const depsLocator = { get: getSensorManager };

    return expectSaga(DownloadSaga.tryDownloadForSensor, { payload: { sensorId: '1' } })
      .provide([[getContext(DEPENDENCY.LOCATOR), depsLocator]])
      .withReducer(DownloadReducer)
      .hasFinalState({
        downloadingById: { 1: false },
        passiveDownloadEnabled: false,
        enabled: false,
      })
      .run();
  });
  it('downloadTemperatures', () => {
    const sensorManager = { getSensors: () => [{ id: '1' }, { id: '2' }] };
    const getSensorManager = () => sensorManager;
    const depsLocator = { get: getSensorManager };

    return expectSaga(DownloadSaga.downloadTemperatures, { payload: { sensorId: '1' } })
      .provide([[getContext(DEPENDENCY.LOCATOR), depsLocator]])
      .run()
      .then(({ effects }) => {
        expect(effects.put.length).toEqual(2);
        expect(effects.put[0]).toEqual(put(DownloadAction.tryPassiveDownloadForSensor('1')));
        expect(effects.put[1]).toEqual(put(DownloadAction.tryPassiveDownloadForSensor('2')));
      });
  });
  it('watchPassiveDownloading: starts a passive download', () => {
    const sensorManager = { getSensors: () => [{ id: '1' }, { id: '2' }] };
    const getSensorManager = () => sensorManager;
    const depsLocator = { get: getSensorManager };

    return expectSaga(DownloadSaga.watchPassiveDownloading)
      .provide([[getContext(DEPENDENCY.LOCATOR), depsLocator]])
      .dispatch(DownloadAction.passiveDownloadingStart())
      .call.fn(DownloadSaga.startPassiveDownloading)
      .silentRun(500);
  });
  it('watchPassiveDownloading: stops a passive download', () => {
    const sensorManager = { getSensors: () => [{ id: '1' }, { id: '2' }] };
    const getSensorManager = () => sensorManager;
    const depsLocator = { get: getSensorManager };

    return expectSaga(DownloadSaga.watchPassiveDownloading)
      .provide([[getContext(DEPENDENCY.LOCATOR), depsLocator]])
      .dispatch(DownloadAction.passiveDownloadingStart())
      .call.fn(DownloadSaga.startPassiveDownloading)
      .dispatch(DownloadAction.passiveDownloadingStop())
      .run();
  });
});
