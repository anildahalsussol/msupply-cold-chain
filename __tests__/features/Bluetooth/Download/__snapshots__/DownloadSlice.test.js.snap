// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DownloadActions  1`] = `
Array [
  Object {
    "payload": undefined,
    "type": "download/passiveDownloadingStart",
  },
  Object {
    "payload": undefined,
    "type": "download/passiveDownloadingStop",
  },
  Object {
    "payload": Object {
      "sensorId": undefined,
    },
    "type": "download/downloadStart",
  },
  Object {
    "payload": Object {
      "sensorId": undefined,
    },
    "type": "download/downloadComplete",
  },
  Object {
    "payload": Object {
      "sensorId": undefined,
    },
    "type": "download/tryManualDownloadForSensor",
  },
  Object {
    "payload": Object {
      "sensorId": undefined,
    },
    "type": "download/manualDownloadForSensorSuccess",
  },
  Object {
    "payload": Object {
      "sensorId": undefined,
    },
    "type": "download/manualDownloadForSensorFail",
  },
  Object {
    "payload": Object {
      "sensorId": undefined,
    },
    "type": "download/tryPassiveDownloadForSensor",
  },
  Object {
    "payload": Object {
      "sensorId": undefined,
    },
    "type": "download/passiveDownloadForSensorSuccess",
  },
  Object {
    "payload": Object {
      "sensorId": undefined,
    },
    "type": "download/passiveDownloadForSensorFail",
  },
]
`;
